crates/sui-graphql-rpc/src/context_data/pg_backend.rs

```diff
@@ -0,0 +1,619 @@
+                            query.filter(objects::dsl::owner_type.eq(OwnerType::Address as i16));
+                    }
+                    Some(OwnerType::Object) => {
+                        query = query.filter(objects::dsl::owner_type.eq(OwnerType::Object as i16));
+                    }
+                    None => {
+                        query = query.filter(
+                            objects::dsl::owner_type
+                                .eq(OwnerType::Address as i16)
+                                .or(objects::dsl::owner_type.eq(OwnerType::Object as i16)),
+                        );
+                    }
+                    _ => Err(DbValidationError::InvalidOwnerType)?,
+                }
+            }
+
+            if let Some(object_type) = filter.ty {
+                query = query.filter(objects::dsl::object_type.eq(object_type));
+            }
+        }
+
+        Ok(query)
+    }
+
+    fn multi_get_balances<'a>(address: Vec<u8>) -> PgBalanceQuery<'a> {
+        let query = objects::dsl::objects
+            .group_by(objects::dsl::coin_type)
+            .select((
+                diesel::dsl::sql::<diesel::sql_types::Nullable<diesel::sql_types::BigInt>>(
+                    "CAST(SUM(coin_balance) AS BIGINT)",
+                ),
+                diesel::dsl::sql::<diesel::sql_types::Nullable<diesel::sql_types::BigInt>>(
+                    "COUNT(*)",
+                ),
+                objects::dsl::coin_type,
+            ))
+            .filter(objects::dsl::owner_id.eq(address))
+            .filter(objects::dsl::owner_type.eq(OwnerType::Address as i16))
+            .filter(objects::dsl::coin_type.is_not_null())
+            .into_boxed();
+
+        query
+    }
+
+    fn get_balance<'a>(address: Vec<u8>, coin_type: String) -> PgBalanceQuery<'a> {
+        let query = PgQueryBuilder::multi_get_balances(address);
+        query.filter(objects::dsl::coin_type.eq(coin_type))
+    }
+
+    fn multi_get_checkpoints<'a>(
+        cursor: Option<i64>,
+        descending_order: bool,
+        limit: i64,
+        epoch: Option<i64>,
+    ) -> checkpoints::BoxedQuery<'a, Pg> {
+        let mut query = checkpoints::dsl::checkpoints.into_boxed();
+
+        if let Some(cursor) = cursor {
+            if descending_order {
+                query = query.filter(checkpoints::dsl::sequence_number.lt(cursor));
+            } else {
+                query = query.filter(checkpoints::dsl::sequence_number.gt(cursor));
+            }
+        }
+        if descending_order {
+            query = query.order(checkpoints::dsl::sequence_number.desc());
+        } else {
+            query = query.order(checkpoints::dsl::sequence_number.asc());
+        }
+        if let Some(epoch) = epoch {
+            query = query.filter(checkpoints::dsl::epoch.eq(epoch));
+        }
+        query = query.limit(limit + 1);
+
+        query
+    }
+}
+
+impl GenericQueryBuilder<Pg> for PgQueryBuilder {
```

Comment:
Can we combine this trait impl with the impl block above? i.e. why do we need to define the functions outside of the trait first?

Reply:
you got it, will do